pkg/template/jsonpath_test.go

```diff
@@ -0,0 +1,145 @@
+package template_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/triggers/pkg/template"
+)
+
+var objects = `{"a":"v","c":{"d":"e"},"empty": "","null": null, "number": 42}`
+var arrays = `[{"a": "b"}, {"c": "d"}, {"e": "f"}]`
+
+// Checks that we print JSON strings when the JSONPath selects
+// an array or map value and regular values otherwise
+func TestParseJSONPath(t *testing.T) {
+	var objectBody = fmt.Sprintf(`{"body":%s}`, objects)
+	tests := []struct {
+		name string
+		expr string
+		in   string
+		want string
+	}{{
+		name: "objects",
+		in:   objectBody,
+		expr: "$(body)",
+		// TODO: Do we need to escape backslashes for backwards compat?
+		want: objects,
+	}, {
+		name: "array of objects",
+		in:   fmt.Sprintf(`{"body":%s}`, arrays),
+		expr: "$(body)",
+		want: arrays,
+	}, {
+		name: "array of values",
+		in:   `{"body": ["a", "b", "c"]}`,
+		expr: "$(body)",
+		want: `["a", "b", "c"]`,
+	}, {
+		name: "string values",
+		in:   objectBody,
+		expr: "$(body.a)",
+		want: "v",
+	}, {
+		name: "string values",
+		in:   objectBody,
+		expr: "$(body.a)",
+		want: "v",
+	}, {
+		name: "empty string",
+		in:   objectBody,
+		expr: "$(body.empty)",
+		want: "",
+	}, {
+		name: "numbers",
+		in:   objectBody,
+		expr: "$(body.number)",
+		want: "42",
+	}, {
+		name: "booleans",
+		in:   `{"body": {"bool": true}}`,
+		expr: "$(body.bool)",
+		want: "true",
+	}, {
+		name: "null values",
+		in:   objectBody,
+		expr: "$(body.null)",
+		want: "null",
+	}}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var data interface{}
+			err := json.Unmarshal([]byte(tt.in), &data)
+			if err != nil {
+				t.Errorf("Could not unmarshall body : %q", err)
+			}
+			got, err := template.ParseJSONPath(data, tt.expr)
+			if err != nil {
+				t.Errorf("ParseJSONPath() error = %v", err)
+				return
+			}
+
+			if diff := cmp.Diff(got, strings.Replace(tt.want, " ", "", -1)); diff != "" {
+				t.Errorf("ParseJSONPath() -got,+want: %s", diff)
+			}
+		})
+	}
+}
+
+func TestParseJSONPath_Error(t *testing.T) {
+	testJSON := `{"body": {"key": "val"}}`
+	invalidExprs := []string{
+		"$({.hello)",
+		"$(+12.3.0)",
+		"$([1)",
+		"$(body",
+		"body)",
+		"body",
+	}
+	var data interface{}
+	err := json.Unmarshal([]byte(testJSON), &data)
+	if err != nil {
+		t.Errorf("Could not unmarshall body : %q", err)
+		return
+	}
+
+	for _, expr := range invalidExprs {
+		t.Run(expr, func(t *testing.T) {
+			got, err := template.ParseJSONPath(data, expr)
+			if err == nil {
+				t.Errorf("ParseJSONPath() did not return expected error; got = %v", got)
+			}
+		})
+	}
+}
+
+func TestTektonJSONPathExpression(t *testing.T) {
+	tests := []struct {
+		expr    string
+		want    string
+		wantErr bool
```

Comment:
i like how the above tests are separate for the error case and the success case, could we do that here as well?

Reply:
done!